app/models/series.rb

```diff
@@ -156,37 +146,12 @@ def allpseuds
     works.collect(&:pseuds).flatten.compact.uniq.sort
   end
 
-  # return list of users on this series
-  def owners
-    self.authors.collect(&:user)
-  end
-
-  # Virtual attribute for pseuds
-  def author_attributes=(attributes)
-    selected_pseuds = Pseud.find(attributes[:ids])
-    (self.authors ||= []) << selected_pseuds
-    # if current user has selected different pseuds
-    current_user = User.current_user
-    if current_user.is_a? User
-      self.authors_to_remove = current_user.pseuds & (self.pseuds - selected_pseuds)
-    end
-    self.authors << Pseud.find(attributes[:ambiguous_pseuds]) if attributes[:ambiguous_pseuds]
-    if !attributes[:byline].blank?
-      results = Pseud.parse_bylines(attributes[:byline], keep_ambiguous: true)
-      self.authors << results[:pseuds]
-      self.invalid_pseuds = results[:invalid_pseuds]
-      self.ambiguous_pseuds = results[:ambiguous_pseuds]
-    end
-    self.authors.flatten!
-    self.authors.uniq!
-  end
-
   # Remove a user as an author of this series
   def remove_author(author_to_remove)
-    pseuds_with_author_removed = self.pseuds - author_to_remove.pseuds
+    pseuds_with_author_removed = pseuds.where.not(user_id: author_to_remove.id)
     raise Exception.new("Sorry, we can't remove all authors of a series.") if pseuds_with_author_removed.empty?
-    Series.transaction do
-      self.pseuds = pseuds_with_author_removed
+    transaction do
+      creatorships.where(pseud: author_to_remove.pseuds).destroy_all
       authored_works_in_series = (author_to_remove.works & self.works)
```

Comment:
Does this overlap with the `remove_from_children` callback in Creatorship? Is it because we want to use a transaction in this case?

I have a similar question for `Work.remove_author`.

Reply:
There's one case that the creatorship callbacks in this PR don't handle at the moment: when there's a co-created work that has a chapter with only one creator, and that creator tries to remove themselves from the work (or from one of the work's series). The way that it works on Beta, the work doesn't complain about an attempt to remove chapter's last creator, but instead sets the chapter's creators equal to the work's (minus the user removing themselves).

That's what the two `remove_author` functions are for: `Work#remove_author` adds the other co-creators to the sole-authored chapters, and `Series#remove_author` calls `Work#remove_author` to handle any works in the series that have tricky chapter creatorships.

I added comments on the two functions to make it clear, refactored the code a little, and added some more tests to make sure that the code is covered.

(If you think it'd be cleaner, I can try changing the `check_not_last` callback to behave differently on chapters -- specifically, copy the work's creatorships at the time the last chapter creatorship is deleted. The downside would be that in some cases where a user has multiple pseuds on the same work, it'd end up saving creatorships and then immediately destroying them, which is kind of wasteful.)

Reply:
Thanks for the explanation! Using `#remove_author` for this case sounds good to me, you don't have to modify `check_not_last`.

We just need to remember to always use `#remove_author` when we need to edit works or series.

For chapters, we don't need `#remove_authors`, is it because removing chapter co-creators doesn't affect works or series, and `check_no_creators` blocks if there are no creators left?

Reply:
Right, the error message is handled elsewhere and it doesn't trigger any creatorship changes for other objects.